### DISCUSSION TITLE
tinyusb vendor class device + stm32g4 + freeRTOS problem, help/explanation needed

### DISCUSSION THREAD

**Discussion by pm-daniel** [hathach/tinyusb#816] [Q&A] [answered]
Hi, 
I've been struggling with getting tinyUSB with freeRTOS to work in my project (the same configuration of tinyUSB [device descriptors, config, and small modification of fsdev source files] worked flawlessly without RTOS). 
My device shows in Device Manager with the status "USB descriptor request failed" 

During my attempts, I concluded that:
- USB device task runs only once (at the very beginning, just after the scheduler is started) then the status of the task remains "blocked"
- task is blocked due to task stack overflow which happens during `tud_task()` 
- after more inspection, I discovered that the very function that causes `vApplicationStackOverflowHook` is `osal_queue_receive` (which is basically a wrapper for `xQueueReceive`)
- status of my USB device task is changed to "blocked" during the `vTaskPlaceOnEventList` function, and then on the second run of the infinite loop (inside `xQueueReceive`) `vApplicationStackOverflowHook` is called and USB device task is never called again

Any ideas or help will be very appreciated!

I'm attaching my RTOS config files, tinyUSB files, and cubeMX .ioc file here:
[additional_files.zip](https://github.com/hathach/tinyusb/files/6411309/additional_files.zip)

**Comment by pm-daniel** [comment]
Apparently, it turns out that `vApplicationStackOverflowHook` was called by another task, after fixing that tinyUSB still doesn't work.
I think there is some problem with USB Queue (the thread is blocked and waiting for anything in Queue to show up - and that does not happen)

**Comment by hathach** [comment, marked as answer]
Have you tried to run one of the _freertos stock example to see if that works for you. If it does, maybe it is good base to start/compare with.

**Reply by pm-daniel** [reply]
Yes, I had a look on _freertos examples (especially CDC+msc one) but they seems to use manually written freeRTOS config that uses freeRTOS with static allocation. Our project is (maybe unfortunately) build using cubeMX - and because of that it is fixed that freeRTOS uses dynamic allocation. I believe it's a simple error in configuration and/or misunderstanding how tinyUSB queue works :/
Just to be clear - same tinyUSB config works perfectly fine without freeRTOS

**Reply by hathach** [reply]
you can enable the FreeRTOS static allocation that is enough for TinyUSB to run, the rest dynamic or not is totally up to your application. From what I see, it is probably freeRTOS config/setup within your project. I don't use cubeMX, so I couldn't help you much.

**Reply by pm-daniel** [reply]
I enabled static allocation but that didn't help either :/  Perhaps I should just ditch cubeMX for freeRTOS configuration, import it into project and configure manually

**Reply by hathach** [reply]
Generated or manual config file is not really matter. You should look at the conifg contents and compared it with the stock example which is known good

**Reply by pm-daniel** [reply]
Indeed, you're so right - after dozen hours of tedious search through my code I realized that I'm retarded - I forget about redirecting USB IRQ to tinyUSB stack - soo apparently two lines of code fixed all my problems. At least I will remember about that forever rotfl

**Reply by hathach** [reply]
I am glad you figure out, it is common issue people have. And that is why I always ask people to try out and start with stock example.

**Reply by ddjangl** [reply]
Just gonna slide in here and follow up. I am doing a pretty similar project except im trying to run the dual cdc ports with freertos. I know that I have re-routed the interrupts to the tud_int_handler() the problem is that WWDG_IRQHandler() gets called at some point and stalls my code out. 

I can't build one of your example projects because I don't have any of the dev boards to test it out on. I am using a STM32G491xx but it isn't utilizing UART to do the USB through the ST-LINK. The device shows up in my device manager but it definitely is not enumerating.
